modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/CacheObjectBinaryProcessorImpl.java

```diff
@@ -535,8 +535,16 @@ public GridBinaryMarshaller marshaller() {
 
             BinaryMetadata mergedMeta = BinaryUtils.mergeMetadata(oldMeta, newMeta0, changedSchemas);
 
-            if (oldMeta != null && mergedMeta == oldMeta && metaHolder.pendingVersion() == metaHolder.acceptedVersion())
-                return; // Safe to use existing schemas.
+            if (mergedMeta == oldMeta) {
+                // Metadata locally is up-to-date. Waiting for updating metadata in an entire cluster, if necessary.
+                if (metaHolder.pendingVersion() == metaHolder.acceptedVersion()) {
```

Comment:
This condition is incorrect: if pending version and accepted version are the same it means metadata is stable, no on-going updates in the cluster, so thread doesn't need to wait for anything and is allowed to proceed right away.

Reply:
I agree it should be vice-versa condition. Thank you for the remark.